File: Lib/FlyCapture2/src/BusEventsEx_VB.Net/BusEventsEx.vb
'=============================================================================
' Copyright © 2017 FLIR Integrated Imaging Solutions, Inc. All Rights Reserved.
'
' This software Is the confidential And proprietary information of FLIR
' Integrated Imaging Solutions, Inc. ("Confidential Information"). You
' shall Not disclose such Confidential Information And shall use it only in
' accordance with the terms of the license agreement you entered into
' with FLIR Integrated Imaging Solutions, Inc. (FLIR).
'
' FLIR MAKES NO REPRESENTATIONS Or WARRANTIES ABOUT THE SUITABILITY OF THE
' SOFTWARE, EITHER EXPRESSED Or IMPLIED, INCLUDING, BUT Not LIMITED TO, THE
' IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
' PURPOSE, Or NON-INFRINGEMENT. FLIR SHALL Not BE LIABLE FOR ANY DAMAGES
' SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING Or DISTRIBUTING
' THIS SOFTWARE Or ITS DERIVATIVES.
'=============================================================================
' =============================================================================
' $Id: BusEventsEx.vb 316528 2017-02-22 00:03:53Z alin $
' =============================================================================

Imports System
Imports System.Text
Imports System.Collections
Imports FlyCapture2Managed

Namespace BusEventsEx_VB
    Class Program

        Shared Sub PrintBuildInfo()
            Dim version As FC2Version = ManagedUtilities.libraryVersion
            Dim newStr As StringBuilder = New StringBuilder()
            newStr.AppendFormat("FlyCapture2 library version: {0}.{1}.{2}.{3}" & vbNewLine, _
                                version.major, version.minor, version.type, version.build)
            Console.WriteLine(newStr)
        End Sub

        Sub OnBusReset(ByVal ptr As IntPtr, ByVal serialNumber As UInteger)
            Console.WriteLine("{0} - *** BUS RESET ***", DateTime.Now.ToString())
        End Sub

        Sub OnBusArrival(ByVal ptr As IntPtr, ByVal serialNumber As UInteger)
            Console.WriteLine("{0} - *** BUS ARRIVAL ({1})***", DateTime.Now.ToString(), serialNumber)
        End Sub

        Sub OnBusRemoval(ByVal ptr As IntPtr, ByVal serialNumber As UInteger)
            Console.WriteLine("{0} - *** BUS REMOVAL ({1})***", DateTime.Now.ToString(), serialNumber)
        End Sub

        Sub BusResetLoop()

            Dim busMgr As ManagedBusManager = New ManagedBusManager()
            Dim callbackHandles As List(Of IntPtr) = New List(Of IntPtr)()

            ' Register bus events
            Dim busResetHandle As IntPtr = busMgr.RegisterCallback(AddressOf OnBusReset, ManagedCallbackType.BusReset, IntPtr.Zero)
            Dim busArrivalHandle As IntPtr = busMgr.RegisterCallback(AddressOf OnBusArrival, ManagedCallbackType.Arrival, IntPtr.Zero)
            Dim busRemovalHandle As IntPtr = busMgr.RegisterCallback(AddressOf OnBusRemoval, ManagedCallbackType.Removal, IntPtr.Zero)

            callbackHandles.Add(busResetHandle)
            callbackHandles.Add(busArrivalHandle)
            callbackHandles.Add(busRemovalHandle)

            ' Prevent exit if CTL+C is pressed.
            Console.TreatControlCAsInput = True

            Console.WriteLine("Press any key to exit..." & vbNewLine)
            Dim cki As ConsoleKeyInfo = Console.ReadKey()

            ' Unregister bus events
            For Each currHandle As IntPtr In callbackHandles
                busMgr.UnregisterCallback(currHandle)
            Next
        End Sub

        Shared Sub Main()
            PrintBuildInfo()

            Dim program As Program = New Program()
            program.BusResetLoop()

            Console.WriteLine("Done! Press enter to exit...")
            Console.ReadLine()
        End Sub

    End Class
End Namespace

File: Lib/FlyCapture2/src/BusEventsEx/BusEventsEx.cpp
//=============================================================================
// Copyright © 2017 FLIR Integrated Imaging Solutions, Inc. All Rights Reserved.
//
// This software is the confidential and proprietary information of FLIR
// Integrated Imaging Solutions, Inc. ("Confidential Information"). You
// shall not disclose such Confidential Information and shall use it only in
// accordance with the terms of the license agreement you entered into
// with FLIR Integrated Imaging Solutions, Inc. (FLIR).
//
// FLIR MAKES NO REPRESENTATIONS OR WARRANTIES ABOUT THE SUITABILITY OF THE
// SOFTWARE, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
// PURPOSE, OR NON-INFRINGEMENT. FLIR SHALL NOT BE LIABLE FOR ANY DAMAGES
// SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING
// THIS SOFTWARE OR ITS DERIVATIVES.
//=============================================================================
//=============================================================================
// $Id: BusEventsEx.cpp 316528 2017-02-22 00:03:53Z alin $
//=============================================================================

#include "stdafx.h"

#ifdef LINUX
#include <unistd.h>
#endif

#include "FlyCapture2.h"
#include <ctime>
#include <iostream>
#include <sstream>

using namespace FlyCapture2;
using namespace std;

void PrintBuildInfo()
{
    FC2Version fc2Version;
    Utilities::GetLibraryVersion(&fc2Version);

    ostringstream version;
    version << "FlyCapture2 library version: " << fc2Version.major << "."
            << fc2Version.minor << "." << fc2Version.type << "."
            << fc2Version.build;
    cout << version.str() << endl;

    ostringstream timeStamp;
    timeStamp << "Application build date: " << __DATE__ << " " << __TIME__;
    cout << timeStamp.str() << endl << endl;
}

void PrintError(Error error) { error.PrintErrorTrace(); }

std::string GetCurrentTimeString()
{
    time_t rawtime;
    struct tm *timeinfo;
    time(&rawtime);
    timeinfo = localtime(&rawtime);

    std::ostringstream formatTime;
    formatTime << (timeinfo->tm_year + 1900) << "-" << (timeinfo->tm_mon + 1)
               << "-" << (timeinfo->tm_mday) << " " << (timeinfo->tm_hour)
               << ":" << (timeinfo->tm_min) << ":" << (timeinfo->tm_sec);

    return formatTime.str();
}

void OnBusReset(void *pParam, unsigned int serialNumber)
{
    std::cout << GetCurrentTimeString() << " - *** BUS RESET ***" << std::endl;
}

void OnBusArrival(void *pParam, unsigned int serialNumber)
{
    std::cout << GetCurrentTimeString() << " - *** BUS ARRIVAL ("
              << serialNumber << ") ***" << std::endl;
}

void OnBusRemoval(void *pParam, unsigned int serialNumber)
{
    std::cout << GetCurrentTimeString() << " - *** BUS REMOVAL ("
              << serialNumber << ") ***" << std::endl;
}

int BusResetLoop()
{
    BusManager busMgr;

    FlyCapture2::CallbackHandle m_resetHandle;
    FlyCapture2::CallbackHandle m_arrivalHandle;
    FlyCapture2::CallbackHandle m_removalHandle;

    // Register bus events
    Error error;
    error = busMgr.RegisterCallback(&OnBusReset, BUS_RESET, NULL, &m_resetHandle);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    error = busMgr.RegisterCallback(&OnBusArrival, ARRIVAL, NULL, &m_arrivalHandle);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    error = busMgr.RegisterCallback(&OnBusRemoval, REMOVAL, NULL, &m_removalHandle);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Prevent exit if CTL+C is pressed.
    std::cout << "Press any key to exit..." << std::endl;
    std::cin.ignore();

    // Unregister bus events
    error = busMgr.UnregisterCallback(m_resetHandle);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    error = busMgr.UnregisterCallback(m_arrivalHandle);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    error = busMgr.UnregisterCallback(m_removalHandle);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    return 0;
}

int main(int /*argc*/, char ** /*argv*/)
{
    PrintBuildInfo();

    // Start bus reset loop
    int retValue = BusResetLoop();
    if (retValue != 0)
    {
        return -1;
    }

    cout << "Done! Press Enter to exit..." << endl;
    cin.ignore();

    return 0;
}
